src/core/reactor.cc

```diff
@@ -1921,6 +1925,147 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
+            sigemptyset(&mask_signals);
+            r = ::posix_spawnattr_setsigmask(&attr, &mask_signals);
+            throw_system_error_on(r != 0, "posix_spawnattr_setsigmask");
+            r = ::posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGDEF | POSIX_SPAWN_SETSIGMASK);
+            throw_system_error_on(r != 0, "posix_spawnattr_setflags");
+
+            return _thread_pool->submit<syscall_result<int>>([&child_pid, &pathname, &actions, &attr, &argv, &env] {
+                std::vector<const char*> argvp;
+                std::transform(argv.cbegin(), argv.cend(), std::back_inserter(argvp),
+                               [](auto& s) { return s.c_str(); });
+                argvp.push_back(nullptr);
+                std::vector<const char*> envp;
+                std::transform(env.cbegin(), env.cend(), std::back_inserter(envp),
+                               [](auto& s) { return s.c_str(); });
+                envp.push_back(nullptr);
+                return wrap_syscall<int>(::posix_spawn(&child_pid, pathname.c_str(), &actions, &attr,
+                                                       const_cast<char* const *>(argvp.data()),
+                                                       const_cast<char* const *>(envp.data())));
+            }).finally([&actions, &attr] {
+                posix_spawn_file_actions_destroy(&actions);
+                posix_spawnattr_destroy(&attr);
+            });
+        }).then([&child_pid, &cout_pipe, &cerr_pipe] (syscall_result<int> ret) {
+            ::close(cout_pipe[1]);
+            ::close(cerr_pipe[1]);
+            if (ret.result != 0) {
+                ::close(cout_pipe[0]);
+                ::close(cerr_pipe[0]);
+                ret.throw_if_error();
+            }
+            return make_ready_future<std::tuple<pid_t, int, int>>(child_pid, cout_pipe[0], cerr_pipe[0]);
+        });
+    });
+}
+
+static const std::chrono::milliseconds wait_pid_poll_period(200);
+
+future<int> reactor::waitpid(pid_t pid) {
+    return _thread_pool->submit<syscall_result<int>>([pid] {
+        return wrap_syscall<int>(syscall(SYS_pidfd_open, pid, O_NONBLOCK));
+    }).then([pid, this] (syscall_result<int> pidfd) {
+        if (pidfd.result == -1) {
+            // pidfd_open() was introduced in linux 5.3, so the pidfd.error could be ENOSYS on
+            // older kernels. But it could be other error like EMFILE or ENFILE. anyway, we
+            // should always waitpid().
+            return do_with(int{}, [pid, this](int& wstatus) {
+                return repeat_until_value([pid, this, &wstatus] {
+                    return _thread_pool->submit<syscall_result<pid_t>>([pid, &wstatus] {
+                        return wrap_syscall<pid_t>(::waitpid(pid, &wstatus, WNOHANG));
+                    }).then([&wstatus] (syscall_result<pid_t> ret) {
+                        if (ret.result == 0) {
+                            return ::seastar::sleep(wait_pid_poll_period).then([] {
```

Comment:
This could start lower and increase, so that tiny forks are fast. But not needed in the first iteration.

Reply:
added an incremental backoff timeout machinery.